ipa-core/src/protocol/ipa_prf/malicious_security/quadratic_proofs.rs

```diff
@@ -0,0 +1,889 @@
+
+        // amount of polynomials in `share`
+        let s = share.0.len() / r_f;
+
+        let mut share_new_0 = vec![F::ONE; s];
+        let mut share_new_1 = vec![F::ONE; s];
+
+        // precompute Lagrange base for `r`
+        let mut base_r = vec![vec![F::ONE; r_f]; 1usize];
+        compute_lagrange_base(
+            std::slice::from_ref(r),
+            &e_p[0..r_f],
+            &denominators,
+            &mut base_r,
+        );
+
+        // `share` is split into chunks of size `recursion_factor`
+        // to represent the polynomials
+        // which are represented by `recursion_factor` many values in `share`
+        // we then Lagrange evaluate it on the next `random_point` to compute new elements in `share`
+        share_new_0
+            .iter_mut()
+            .zip(share.0.chunks(r_f))
+            .for_each(|(u, chunk)| {
+                lagrange_evaluation_precomputed(chunk, &base_r, std::slice::from_mut(u));
+            });
+        share_new_1
+            .iter_mut()
+            .zip(share.1.chunks(r_f))
+            .for_each(|(u, chunk)| {
+                lagrange_evaluation_precomputed(chunk, &base_r, std::slice::from_mut(u));
+            });
+        // reassign new `share` for next iteration
+        *share.0 = share_new_0;
+        *share.1 = share_new_1;
+        share.0.truncate(s);
+        share.1.truncate(s);
+    }
+
+    debug_assert!(share.0.len() < r_f);
+
+    // final shares, compute output
+    let mut output_0 = F::ONE;
+    let mut output_1 = F::ONE;
+
+    let (final_share_0, final_share_1) = {
+        let mut vec_0 = Vec::<F>::with_capacity(r_f);
+        let mut vec_1 = Vec::<F>::with_capacity(r_f);
+        // add prss mask
+        vec_0.push(*mask);
+        vec_1.push(*mask);
+        // copy rest
+        vec_0.extend(&*share.0);
+        vec_1.extend(&*share.1);
+        // fill u and v with zeros such that it is a multiple of the recursion factor
+        vec_0.extend(std::iter::repeat(F::ZERO).take(vec_0.len() % r_f));
+        vec_1.extend(std::iter::repeat(F::ZERO).take(vec_1.len() % r_f));
+        (vec_0, vec_1)
+    };
+
+    // precompute Lagrange base for `r`
+    let mut base_r = vec![vec![F::ONE; r_f]; 1usize];
+    compute_lagrange_base(
+        std::slice::from_ref(r_p.last().unwrap()),
+        &e_p[0..r_f],
+        &denominators,
+        &mut base_r,
+    );
+
+    lagrange_evaluation_precomputed(&final_share_0, &base_r, std::slice::from_mut(&mut output_0));
+    lagrange_evaluation_precomputed(&final_share_1, &base_r, std::slice::from_mut(&mut output_1));
+
+    return (output_0, output_1);
+}
+
+/// computes `summand_of_g = p*q` and adds it to `g`
+/// polynomials `p` and `q` are represented by value `u`, `v`
+/// which are the evaluations of `p,q` at `evaluation_points[0..recursion_factor]`
+/// `g` is represented by its evaluation at `evaluation_points`
+fn compute_summand_of_g<F>(u: &[F], v: &[F], r_f: usize, base: &Vec<Vec<F>>, g: &mut [F]) -> ()
```

Comment:
This should take in iterators and return a value.  Having it mutate an argument makes it a lot less flexible.  Also, this requires additional and potentially unnecessary allocations.

Given that `g` is only added to, could you make this function produce a "map" operand?  That is, you would do something like this:

```rust
g = g.into_iter().map(Summand::new(u, v, precomputed).as_f()).collect::<Vec<_>>();
```

or just
```rust
g = Summand::new(u, v, precomputed) + g;
```

Which could implement `Add<Rhs = X> where X: AsRef<F: Field + Invert>` or similar (excuse the abuse of notation).